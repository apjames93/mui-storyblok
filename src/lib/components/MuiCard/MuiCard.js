import React from 'react';
import PropTypes from 'prop-types';
import { Card } from '@material-ui/core';
import Storyblok from '../../utils/Storyblok';
import { renderComponents } from '../../utils/customComponents';
import {
  validComponents,
  dimensionProp,
} from '../../utils/customProps';
import MuiCardActions from './components/MuiCardActions/MuiCardActions';
import MuiCardContent from './components/MuiCardContent/MuiCardContent';
import MuiCardHeader from './components/MuiCardHeader/MuiCardHeader';
import MuiCardMedia from './components/MuiCardMedia/MuiCardMedia';
import MuiVideo from '../MuiVideo/MuiVideo';

const MuiCard = ({
  rootClass,
  raised,
  content,
  elevation,
  square,
  variant,
  height,
  width,
}) => {
  const components = {
    MuiCardActions,
    MuiCardContent,
    MuiCardHeader,
    MuiCardMedia,
    MuiVideo,
  };
  const styles = Storyblok.arrayToMuiStyles(rootClass, { height, width });
  return (
    <Card
      className={styles.root}
      raised={raised}
      variant={variant}
      elevation={+elevation}
      square={square}
    >
      {content.map((component, key) => renderComponents(components, component, key))}
    </Card>
  );
};

export default MuiCard;

MuiCard.propTypes = {
  /**
   * storyblok multiselect of css classes
   * Mui Override or extend the styles applied to the component.
   */
  rootClass: PropTypes.arrayOf(PropTypes.string),
  /**
   * mui prop: true | false
   * If true, the card will use raised styling.
   */
  raised: PropTypes.bool,
  /**
   * It accepts values between 0 and 24 inclusive.
   * Shadow depth, corresponds to dp in the spec.
   */
  elevation: PropTypes.string,
  /**
   * mui prop: true | false
   * If true, rounded corners are disabled.
   */
  square: PropTypes.bool,
  /**
   * mui prop: 'elevation' | 'outlined'
   * The variant to use.
   */
  variant: PropTypes.string,
  /**
   * Height of the card.
   */
  height(props, propName, componentName) {
    return dimensionProp(props, propName, componentName);
  },
  /**
   * Width of the card.
   */
  width(props, propName, componentName) {
    return dimensionProp(props, propName, componentName);
  },

  /**
   * Content passed to render
   * can also render any customCompnent passed in
   * components:  MuiCardActions,
    MuiCardContent,
    MuiCardHeader,
    MuiCardMedia,
   */
  content(props, propName, componentName) {
    const components = [
      'MuiCardActions',
      'MuiCardContent',
      'MuiCardHeader',
      'MuiCardMedia',
    ];
    return validComponents(props, propName, componentName, components);
  },
};

MuiCard.defaultProps = {
  raised: false,
  rootClass: [],
  height: '100%',
  width: '100%',
  content: [],
  elevation: 1,
  square: false,
  variant: 'elevation',
};
