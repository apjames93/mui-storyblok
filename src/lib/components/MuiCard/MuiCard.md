##### MuiCard has the ability to render custom components.

MuiCard With MuiCardActions:

```js
const cardcontent2 = {
    component: 'MuiCardContent',
    content: [{
        component: 'MuiTypography',
        content: [{
            component: 'MuiText',
            content: "This is a card",
        }],
    }],
}

const action2 = {
    component: 'MuiCardActions',
    content: [{
        component: 'MuiIconButtonDownload',
        href: "https://miro.medium.com/max/1200/1*mk1-6aYaf_Bes1E3Imhc0A.jpeg",
        fileName: "cool.jpeg",
        iconButton: [{
        component: 'MuiIconButton',
        icon: [{
            component: 'MuiIcon',
            iconName: 'android',
        }],
      }],
    }]
};

const content2 = [cardcontent2, action2];

  <MuiCard
    content={content2}
  />

```

MuiCard With MuiCardActions, MuiCardMedia, MuiCardHeader, MuiCardContent:

```js
const header = {
        component: 'MuiCardHeader',
        title: 'Card Example with card media and Action',
        content: [{
            component: 'MuiIcon',
            iconName: "star",
        }]
    }

const media = {
        component: 'MuiCardMedia',
        alt: 'alt',
        title: 'title',
        image: 'https://cdn.arstechnica.net/wp-content/uploads/2016/02/5718897981_10faa45ac3_b-640x624.jpg',
    }

const cardcontent = {
    component: 'MuiCardContent',
    content: [{
        component: 'MuiTypography',
        content: [{
            component: 'MuiText',
            content: "whatever dude",
        }],
    }],
}


const action = {
    component: 'MuiCardActions',
    content: [{
        component: 'MuiIconButtonDownload',
        href: "https://miro.medium.com/max/1200/1*mk1-6aYaf_Bes1E3Imhc0A.jpeg",
        fileName: "cool.jpeg",
        iconButton: [{
        component: 'MuiIconButton',
        icon: [{
            component: 'MuiIcon',
            iconName: 'android',
        }],
      }],
    }]
};

const content = [header, media, cardcontent, action];


  <MuiCard
    content={content}
  />

```