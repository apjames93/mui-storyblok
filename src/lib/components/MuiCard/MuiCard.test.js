import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import MuiCard from './MuiCard';

function setup() {
  const props = {
    _editable: '<!--#storyblok#{"name": "MuiCard", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
    content: [{
      _editable: '<!--#storyblok#{"name": "MuiCardActions", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
      component: 'MuiCardActions',
      content: [{
        _editable: '<!--#storyblok#{"name": "MuiButtonRedirect", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
        component: 'MuiButtonRedirect',
        redirectRoute: 'page-welcome',
        button: [{
          _editable: '<!--#storyblok#{"name": "MuiButton", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
          component: 'MuiButton',
          buttonText: 'buttonText',
        }],
      }],
    }, {
      _editable: '<!--#storyblok#{"name": "MuiCardContent", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
      component: 'MuiCardContent',
      content: [{
        _editable: '<!--#storyblok#{"name": "MuiTypography", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
        component: 'MuiTypography',
        content: [{
          _editable: '<!--#storyblok#{"name": "MuiText", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
          component: 'MuiText',
          text: 'text',
        }],
      }],
    }, {
      _editable: '<!--#storyblok#{"name": "MuiCardHeader", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
      component: 'MuiCardHeader',
      title: 'title',
    }, {
      _editable: '<!--#storyblok#{"name": "MuiCardMedia", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
      component: 'MuiCardMedia',
      alt: 'alt',
      title: 'title',
      image: 'img.com',
    }],
  };
  const comp = shallow(<MuiCard {...props} />);
  return { comp, props };
}

describe('<MuiCard />', () => {
  it('renders MuiCard', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });

  test('snapshot', () => {
    const { props } = setup();

    const tree = renderer.create(<MuiCard {...props} />);
    expect(tree).toMatchSnapshot();
  });
});
