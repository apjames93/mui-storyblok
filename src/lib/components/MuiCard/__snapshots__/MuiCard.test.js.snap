// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MuiCard /> snapshot 1`] = `
<div
  className="MuiPaper-root MuiCard-root makeStyles-root-3 MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="MuiCardActions-root makeStyles-root-4 MuiCardActions-spacing"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-root-5 MuiButton-textSecondary MuiButton-fullWidth"
      data-testid="muiButton"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        buttonText
      </span>
    </button>
  </div>
  <div
    className="MuiCardContent-root makeStyles-root-6"
  >
    <div
      className="MuiTypography-root makeStyles-root-7 MuiTypography-body1"
    >
      <span
        className="MuiTypography-root makeStyles-root-8 MuiTypography-body1"
      >
         undefined 
      </span>
    </div>
  </div>
  <div
    className="MuiCardHeader-root makeStyles-root-9"
  >
    <div
      className="MuiCardHeader-avatar"
    />
    <div
      className="MuiCardHeader-content"
    >
      <span
        className="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
      >
        title
      </span>
      <span
        className="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
      >
        
      </span>
    </div>
    <div
      className="MuiCardHeader-action"
    />
  </div>
  <div
    alt="alt"
    className="MuiCardMedia-root makeStyles-root-10"
    style={
      Object {
        "backgroundImage": "url(\\"img.com\\")",
      }
    }
    title="title"
  />
</div>
`;
